File: ruby/206_reverse_linked_list.rb
# 206. Reverse Linked List
#
# Given the head of a singly linked list, reverse the list, and return the reversed list.

class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, next_ = nil)
    @val = val
    @next = next_
  end
end

def reverse_list(head)
  p = nil
  c = head
  n = nil

  until c.nil?
    n = c.next
    c.next = p
    p = c
    c = n
  end

  p
end

l1 = ListNode.new(1, ListNode.new(2, ListNode.new(3, ListNode.new(4, ListNode.new(5)))))
p reverse_list(l1)

l2 = ListNode.new(1, ListNode.new(2))
p reverse_list(l2)

p reverse_list(nil)

File: python/206_reverse_linked_list.py
"""
206. Reverse Linked List
Easy

Given the head of a singly linked list, reverse the list, and return the reversed list.
"""


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        next = None

        while curr is not None:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next

        return prev


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
